### ISSUE TITLE
GraphQL Rest API

### ISSUE THREAD

**Issue by Speedy1991** [pennersr/django-allauth#2114]
Hello!
Our company is working with django-allauth and we use react + Apollo (and react-native) as frontend. We also have wagtail (django cms).
All the communication is going via GraphQL and we want to have a central user administration in our API.

I'm going to implement a Graphene/GraphQL Frontend for django-allauth and would be happy if i can contribute to this awsome project.

First step would be to clone all the form and view logic to GraphQL mutations.
As far i understand, i will need to use the adapter to manage all the frontend logic. But the current adapter is wired to forms.

e.g.: https://github.com/pennersr/django-allauth/blob/master/allauth/account/adapter.py#L217
`def save_user(self, request, user, form, commit=True):`

So i can't just extend the existing Adapter to make things working - but i could change the adapter/service to take clean_data instead of the form, e.g.:
`def save_user(self, request, user, commit=True, **kwargs):`

This will make the adapter not form related and therefore reusable for e.g. the GraphQL frontend.
A form call could be something like:
`save_user(request, user, commit=True, **self.cleaned_data):`

For making "old" things working again, the `save_user` function will get patched like this:
https://github.com/pennersr/django-allauth/blob/master/allauth/account/adapter.py#L224
```
last_name = kwargs.get('last_name')
email = kwargs.get('email')
username = kwargs.get('username')
```

Are there other pitfalls or things i have to look out for?

Some Questions related to the adapter:
- I don't fully understand the ajax related stuff. Is it processing the redirection link from e.g. Google OAuth? When is this method needed/called?
- What is the ajax_response_form needed for?

Things i can contribute:
1. Step
- Patching the adapter
2. Step
- Create GraphQL Mutations for the current views (account)
- Tests for the GraphQL Mutations
3. Step
- Create GraphQL Mutations for the current views (socialaccount)
- Tests for the GraphQL Mutations

Additional dependencies afterwards:
- graphene
- django-graphene

**Comment by Speedy1991** [comment]
After digging deeper in the allauth package i think i don't need to mirror the adapter. I think it will be enough to mirror the views

**Comment by pennersr** [comment]
From an allauth point of view, I do not want to provide graphene support (and adding dependencies like django-graphene etc.) out of the box.  This is better served in a project of its own that builds on top of allauth. If there is anything that complicates things while trying to create such a separate package, then I am definitely interested to discuss if there is anything that can be done here to accomodate.

**Comment by Speedy1991** [comment]
Thanks for the feedback!

I will close this issue + related PR

**Comment by igo** [comment]
@Speedy1991 did your work ended up in some library? Or has everyone do this from scratch?

**Comment by Speedy1991** [comment]
No library yet, but a standalone django app - I'll ask my boss if we can open source it.

The framework was a way to much generic for my needs, so I stripped it completely down, e.g. I didn't need 90% of the forms and views. I also reworked most of the settings and core logic

So my "library" hasn't too much in common with django-allauth except the OAuth-Clients & Providers

IMO:

To be honest, if you only want to support OAuth2Clients it is a way more efficient to copy the OAuth2Client and initialize it with the needed providers on the fly. You can throw away allauth and all dependencies from your project and end up in about 500 lines of code inclusive JWT + graphql
Thats how we do it now

**Comment by Suor** [comment]
@Speedy1991 Can you share that in any form? Gist/github repo even unstructured and with no docs. We are going to do exactly the same. Only need Github/Gitlab login, some data access via GraphQL.

**Comment by karthikdeva** [comment]
@Speedy1991  Were you able to complete the GraphQL implementation?

**Comment by Speedy1991** [comment]
Hey @karthikdeva – really appreciate you bringing this up again!

I looked into Allauth, but it didn’t quite fit my needs. It felt a bit too bulky, tightly coupled to Django views, and the adapters were somewhat unstable. I’m not sure about the current state of the repo, but I did notice there's now a [headless option](https://docs.allauth.org/en/dev/headless/index.html), which looks pretty promising!

From what I knew about allauth, I wouldn’t recommend it for a large project - it comes with a lot of built-in features, which can be great, but often lead to overfitting. For common OAuth providers like Google and Apple, the implementation is quite lean - roughly 100 lines of code per provider—and doesn’t require a 3rd-party library.

That said, Allauth is an awesome package with a ton of functionality, but for my use cases, it’s just a bit too heavy (thousands of lines of code and a 3rd party dependency vs. ~400 for handling Google, Apple, and Meta authentication).
